File: scala/revenj-core/src/main/scala/net/revenj/patterns/DataChangeNotification.scala
package net.revenj.patterns

import monix.reactive.Observable

import scala.concurrent.Future
import scala.reflect.ClassTag

trait DataChangeNotification {

  def notifications: Observable[DataChangeNotification.NotifyInfo]

  def track[T : ClassTag](implicit manifest: ClassTag[T]): Observable[DataChangeNotification.TrackInfo[T]]
}

object DataChangeNotification {
  sealed trait Operation
  object Operation {
    case object Insert extends Operation
    case object Update extends Operation
    case object Change extends Operation
    case object Delete extends Operation
  }

  sealed trait Source
  object Source {
    case object Database extends Source
    case object Local extends Source
  }

  class NotifyInfo(val name: String, val operation: Operation, val source: Source, val uris: IndexedSeq[String])
  class NotifyWith[T](name: String, operation: Operation, source: Source, uris: IndexedSeq[String], val info: T)
    extends NotifyInfo(name, operation, source, uris)
  object NotifyInfo {
    def apply(name: String, operation: Operation, source: Source, uris: IndexedSeq[String]): NotifyInfo = {
      new NotifyInfo(name, operation, source, uris)
    }
    def apply[T <: Identifiable](name: String, operation: Operation, objects: Seq[T]): NotifyWith[Seq[T]] = {
      new NotifyWith(name, operation, DataChangeNotification.Source.Local, objects.map(_.URI).toIndexedSeq, objects)
    }
  }
  case class TrackInfo[T](uris: IndexedSeq[String], result: Function0[Future[IndexedSeq[T]]])

}

File: java/revenj-core/src/main/java/org/revenj/patterns/DataChangeNotification.java
package org.revenj.patterns;

import rx.Observable;

import java.util.List;
import java.util.concurrent.Callable;

public interface DataChangeNotification {

	final class NotifyInfo {
		public final String name;
		public final Operation operation;
		public final Source source;
		public final String[] uris;

		public NotifyInfo(String name, Operation operation, Source source, String[] uris) {
			this.name = name;
			this.operation = operation;
			this.source = source;
			this.uris = uris;
		}
	}

	enum Operation {Insert, Update, Change, Delete}
	enum Source {Database, Local}

	Observable<NotifyInfo> getNotifications();

	final class TrackInfo<T> {
		public final String[] uris;
		public final Callable<List<T>> result;

		public TrackInfo(String[] uris, Callable<List<T>> result) {
			this.uris = uris;
			this.result = result;
		}
	}

	<T> Observable<TrackInfo<T>> track(Class<T> manifest);
}
